File: paleo_workbench/harness/registry.py
"""Action registry (P2-C): the single authority for harness actions.

Registration is explicit and code-owned (actions live in
:mod:`paleo_workbench.harness.actions`); the registry validates specs,
refuses duplicate ids, and derives agent tool schemas from the same
ActionSpec objects used for runtime validation — one source of truth.
"""
from __future__ import annotations

import logging
import threading
from typing import Any

from paleo_workbench.harness.spec import ActionRisk, ActionSpec, validate_action_spec

logger = logging.getLogger(__name__)


class UnknownActionError(LookupError):
    def __init__(self, action_id: str):
        self.action_id = action_id
        super().__init__(f"unknown harness action {action_id!r}")


class DuplicateActionError(ValueError):
    def __init__(self, action_id: str):
        self.action_id = action_id
        super().__init__(f"harness action {action_id!r} already registered")


class InvalidActionSpecError(ValueError):
    def __init__(self, action_id: str, problems: list[str]):
        self.problems = problems
        super().__init__(f"invalid action spec {action_id!r}: {'; '.join(problems)}")


class ActionRegistry:
    def __init__(self) -> None:
        self._lock = threading.RLock()
        self._actions: dict[str, ActionSpec] = {}

    # ------------------------------------------------------------ register --
    def register(self, spec: ActionSpec, *, replace: bool = False) -> ActionSpec:
        problems = validate_action_spec(spec)
        if problems:
            raise InvalidActionSpecError(spec.action_id, problems)
        if spec.risk is ActionRisk.DESTRUCTIVE:
            # Destructive actions exist in the vocabulary but are refused by
            # the default registry: product policy keeps purge/overwrite out
            # of the agent surface entirely.
            raise InvalidActionSpecError(
                spec.action_id, ["DESTRUCTIVE actions are not installable in the default registry"]
            )
        with self._lock:
            if spec.action_id in self._actions and not replace:
                raise DuplicateActionError(spec.action_id)
            self._actions[spec.action_id] = spec
        return spec

    def unregister(self, action_id: str) -> bool:
        with self._lock:
            return self._actions.pop(action_id, None) is not None

    # -------------------------------------------------------------- lookup --
    def get(self, action_id: str) -> ActionSpec:
        spec = self.find(action_id)
        if spec is None:
            raise UnknownActionError(action_id)
        return spec

    def find(self, action_id: str) -> ActionSpec | None:
        with self._lock:
            return self._actions.get(action_id)

    def specs(self, domain: str | None = None) -> list[ActionSpec]:
        with self._lock:
            specs = list(self._actions.values())
        if domain is not None:
            specs = [s for s in specs if s.domain == domain]
        return sorted(specs, key=lambda s: s.action_id)

    def domains(self) -> list[str]:
        return sorted({s.domain for s in self.specs()})

    def __len__(self) -> int:
        with self._lock:
            return len(self._actions)

    # ---------------------------------------------------------- tool layer --
    def tool_schemas(self, domain: str | None = None) -> list[dict[str, Any]]:
        """Agent-facing tool definitions derived from the specs."""
        return [spec.tool_schema() for spec in self.specs(domain)]

    def inventory(self) -> list[dict[str, Any]]:
        return [spec.to_dict() for spec in self.specs()]


_GLOBAL_REGISTRY: ActionRegistry | None = None
_REGISTRY_LOCK = threading.Lock()


def get_action_registry() -> ActionRegistry:
    global _GLOBAL_REGISTRY
    with _REGISTRY_LOCK:
        if _GLOBAL_REGISTRY is None:
            registry = ActionRegistry()
            from paleo_workbench.harness.actions import register_all

            register_all(registry)
            _GLOBAL_REGISTRY = registry
        return _GLOBAL_REGISTRY


def set_action_registry(registry: ActionRegistry | None) -> None:
    global _GLOBAL_REGISTRY
    with _REGISTRY_LOCK:
        _GLOBAL_REGISTRY = registry

File: libs/closure_agent/src/registry.cpp
// registry.cpp — ActionRegistry behaviour with message parity against the
// frozen Python harness (harness/registry.py).
#include <pwb/closure_agent/registry.hpp>

#include <algorithm>

#include <pwb/providers/errors.hpp>
#include <pwb/providers/schema.hpp>

namespace pwb::closure_agent {

UnknownActionError::UnknownActionError(std::string action_id)
    : std::runtime_error("unknown harness action " +
                         pwb::providers::python_repr(action_id)),
      action_id_(std::move(action_id)) {}

DuplicateActionError::DuplicateActionError(std::string action_id)
    : std::runtime_error("harness action " +
                         pwb::providers::python_repr(action_id) +
                         " already registered"),
      action_id_(std::move(action_id)) {}

InvalidActionSpecError::InvalidActionSpecError(std::string action_id,
                                               std::vector<std::string> problems)
    : std::runtime_error("invalid action spec " +
                         pwb::providers::python_repr(action_id) + ": " +
                         [&] {
                             std::string joined;
                             for (std::size_t i = 0; i < problems.size(); ++i) {
                                 if (i) joined += "; ";
                                 joined += problems[i];
                             }
                             return joined;
                         }()),
      action_id_(std::move(action_id)),
      problems_(std::move(problems)) {}

const ActionSpec& ActionRegistry::register_spec(const ActionSpec& spec,
                                                bool replace) {
    auto problems = validate_action_spec(spec);
    if (spec.risk == ActionRisk::Destructive) {
        // Destructive actions exist in the vocabulary but are refused by the
        // default registry: product policy keeps purge/overwrite out of the
        // agent surface entirely.
        problems.push_back(
            "DESTRUCTIVE actions are not installable in the default registry");
    }
    if (!problems.empty()) {
        throw InvalidActionSpecError(spec.action_id, std::move(problems));
    }
    std::lock_guard<std::mutex> lock(mutex_);
    const auto it = actions_.find(spec.action_id);
    if (it != actions_.end() && !replace) {
        throw DuplicateActionError(spec.action_id);
    }
    ActionSpec copy = spec;
    return actions_.insert_or_assign(spec.action_id, std::move(copy)).first->second;
}

bool ActionRegistry::unregister(const std::string& action_id) {
    std::lock_guard<std::mutex> lock(mutex_);
    return actions_.erase(action_id) > 0;
}

const ActionSpec& ActionRegistry::get(const std::string& action_id) const {
    const ActionSpec* spec = find(action_id);
    if (spec == nullptr) throw UnknownActionError(action_id);
    return *spec;
}

const ActionSpec* ActionRegistry::find(const std::string& action_id) const {
    std::lock_guard<std::mutex> lock(mutex_);
    const auto it = actions_.find(action_id);
    return it == actions_.end() ? nullptr : &it->second;
}

std::vector<const ActionSpec*> ActionRegistry::specs(
    const std::optional<std::string>& domain) const {
    std::lock_guard<std::mutex> lock(mutex_);
    std::vector<const ActionSpec*> selected;
    // std::map is already ordered by key == sorted by action_id.
    for (const auto& [id, spec] : actions_) {
        if (domain && spec.domain() != *domain) continue;
        selected.push_back(&spec);
    }
    return selected;
}

std::vector<std::string> ActionRegistry::domains() const {
    std::vector<std::string> names;
    for (const auto* spec : specs()) names.push_back(spec->domain());
    std::sort(names.begin(), names.end());
    names.erase(std::unique(names.begin(), names.end()), names.end());
    return names;
}

std::size_t ActionRegistry::size() const {
    std::lock_guard<std::mutex> lock(mutex_);
    return actions_.size();
}

std::vector<Json> ActionRegistry::tool_schemas(
    const std::optional<std::string>& domain) const {
    std::vector<Json> schemas;
    for (const auto* spec : specs(domain)) schemas.push_back(spec->tool_schema());
    return schemas;
}

std::vector<Json> ActionRegistry::inventory() const {
    std::vector<Json> entries;
    for (const auto* spec : specs()) entries.push_back(spec->to_dict());
    return entries;
}

}  // namespace pwb::closure_agent
